preprocessors/openstreetmap/osm_service.py

```diff
@@ -0,0 +1,388 @@
+                intersection_record.append(street_object)
+
+    # Group the street segment by their ids
+    output = {}
+    for obj in intersection_record:
+        street_id = obj["street_id"]
+        if street_id not in output:
+            assert obj["intersection_nodes"] is not None
+            record = {
+                "street_id": obj["street_id"],
+                "street_name": obj["street_name"],
+                "intersection_nodes": obj["intersection_nodes"],
+            }
+            output[street_id] = record
+        else:
+            existing_record = output[street_id]
+            existing_intersection_nodes = existing_record["intersection_nodes"]
+            assert existing_intersection_nodes is not None
+            new_intersection_nodes = obj["intersection_nodes"]
+            q = new_intersection_nodes
+            assert q is not None
+            merged_intersection_nodes = existing_intersection_nodes + q
+            existing_record["intersection_nodes"] = merged_intersection_nodes
+            output[street_id] = existing_record
+    intersection_record_updated = list(output.values())
+    # Keep a unique set of intersections under each street segment (i.e.
+    # street id)
+    for obj in range(len(intersection_record_updated)):
+        unique_set = []
+        inter_sets = intersection_record_updated[obj]["intersection_nodes"]
+        for item in range(len(inter_sets)):
+            if inter_sets[item] not in unique_set:
+                unique_set.append(inter_sets[item])
+        intersection_record_updated[obj]["intersection_nodes"] = unique_set
+    return (intersection_record_updated)
+
+
+def allot_intersection(processed_OSM_data, inters_rec_up
+                       ):  # iterate & indicate common nodes
+    processed_OSM_data1 = deepcopy(processed_OSM_data)
+    inters = inters_rec_up
+    for obj in range(len(processed_OSM_data1)):
+        name1 = processed_OSM_data1[obj]["street_name"]
+        id1 = processed_OSM_data1[obj]["street_id"]
+        nodes = processed_OSM_data1[obj]["nodes"]
+        for i in range(len(nodes)):
+            for objs in range(len(inters)):
+                name2 = inters[objs]["street_name"]
+                id2 = inters[objs]["street_id"]
+                intersection_nodes = inters[objs]["intersection_nodes"]
+                for items in range(len(intersection_nodes)):
+                    if id1 != id2:  # compare unique street segment only
+
+                        # check if a node represents an intersection
+                        if nodes[i] == intersection_nodes[items]:
+                            nodes[i]["cat"] = "intersection"
+                            f = nodes[i]
+                            f["name"] = f"{name1}{id1} intersects {name2}{id2}"
+
+    return processed_OSM_data1
+
+
+def get_amenities(bbox_coordinates):
+    # Send request OSM to get amenities which are part of point of interest
+    # (POIs)
+    api = overpy.Overpass()
```

Comment:
Overpass API is part of OSM?

Reply:
Yes, OSM uses Overpass API.
https://wiki.openstreetmap.org/wiki/Overpass_API